staging/src/k8s.io/dynamic-resource-allocation/resourceslice/noderesources.go

```diff
@@ -0,0 +1,392 @@
+		informer.Run(ctx.Done())
+	}()
+	for !handler.HasSynced() {
+		select {
+		case <-time.After(time.Second):
+		case <-ctx.Done():
+			return
+		}
+	}
+	logger.V(3).Info("ResourceSlice informer has synced")
+
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	key, shutdown := c.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	// Panics are caught and treated like errors.
+	var err error
+	func() {
+		defer func() {
+			if r := recover(); r != nil {
+				err = fmt.Errorf("internal error: %v", r)
+			}
+		}()
+		err = c.sync(ctx)
+	}()
+
+	if err != nil {
+		// TODO (https://github.com/kubernetes/enhancements/issues/3077): contextual logging in utilruntime
+		utilruntime.HandleError(fmt.Errorf("processing ResourceSlice objects: %v", err))
+		c.queue.AddRateLimited(key)
+
+		// Return without removing the work item from the queue.
+		// It will be retried.
+		return true
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *Controller) sync(ctx context.Context) error {
+	logger := klog.FromContext(ctx)
+
+	// Gather information about the actual and desired state.
+	slices := c.sliceStore.List()
+	var resources *Resources
+	c.mutex.RLock()
+	resources = c.resources
+	c.mutex.RUnlock()
+
+	// Resources that are not yet stored in any slice need to be published.
+	// Here we track the indices of any resources that are already stored.
+	storedResourceIndices := sets.New[int]()
+
+	// Slices that don't match any driver resource can either be updated (if there
+	// are new driver resources that need to be stored) or they need to be deleted.
+	obsoleteSlices := make([]*resourceapi.ResourceSlice, 0, len(slices))
+
+	// Match slices with resource information.
+	for _, obj := range slices {
+		slice := obj.(*resourceapi.ResourceSlice)
+
+		// TODO: network-attached resources.
+		index := indexOfModel(resources.NodeResources, &slice.ResourceModel)
+		if index >= 0 {
+			storedResourceIndices.Insert(index)
+			continue
+		}
+
+		obsoleteSlices = append(obsoleteSlices, slice)
+	}
+
+	if loggerV := logger.V(6); loggerV.Enabled() {
+		// Dump entire resource information.
+		loggerV.Info("Syncing existing driver resource slices with driver resources", "slices", klog.KObjSlice(slices), "resources", resources)
+	} else {
+		logger.V(5).Info("Syncing existing driver resource slices with driver resources", "slices", klog.KObjSlice(slices), "numResources", len(resources.NodeResources))
+	}
+
+	// Retrieve node object to get UID?
+	// The result gets cached and is expected to not change while
+	// the controller runs.
+	if c.owner.UID == "" && c.owner.APIVersion == "v1" && c.owner.Kind == "Node" {
+		node, err := c.kubeClient.CoreV1().Nodes().Get(ctx, c.owner.Name, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("retrieve node %q: %v", c.owner.Name, err)
+		}
+		// There is only one worker, so no locking needed.
+		c.owner.UID = node.UID
+	}
```

Comment:
The caller in this case is the DRA kubelet plugin. It knows the node name, but not the node UID.

We could make it the responsibility of each DRA kubelet plugin to fetch the node object before instantiating this helper code, but isn't it simpler to do that automatically here?

Reply:
Yes, this is what I was suggesting. I'm fine making it an additional Option to be passed (in addition to NodeName) instead of forcing the caller to supply it.

Reply:
Actually, I think the part I don't like the most is passing in an incomplete object for the owner, and then filling it out here later. Can we instead do this at the point that we are creating the `owner` object, so that a complete object is passed into this function?

Reply:
I added a `NodeUID` option, although I don't think that it'll be used much.